kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java

```diff
@@ -0,0 +1,417 @@
+ *    <li>The registry provides an {@link EndpointResolver}.  The {@link #resolve(String, int, String, boolean)} method accepts
+ * connection metadata (port, SNI etc) and resolves this to a @{@link VirtualClusterBinding}.  This allows
+ * Kroxylicious to determine the destination of any incoming connection.</li>
+ * </ul>
+ * The registry is thread safe for all operations.
+ * <ul>
+ *    <li>virtual cluster registration uses java.util.concurrent features to ensure registration is single threaded.</li>
+ *    <li>virtual cluster de-registration uses java.util.concurrent.atomic to ensure de-registration is single threaded.</li>
+ *    <li>updates to the binding mapping (attached to channel) are made only whilst holding an intrinsic lock on the {@link ListeningChannelRecord}.</li>
+ *    <li>updates to the binding mapping are published safely to readers (i.e. threads calling {@link #resolve(String, int, String, boolean)}).  This relies the
+ *    fact that the binding map uses concurrency safe data structures ({@link ConcurrentHashMap} and the exclusive use of immutable objects within it.</li>
+ * </ul>
+ */
+public class EndpointRegistry implements AutoCloseable, EndpointResolver {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EndpointRegistry.class);
+    private final AtomicBoolean registryClosed = new AtomicBoolean(false);
+
+    interface RoutingKey {
+        RoutingKey NULL_ROUTING_KEY = new NullRoutingKey();
+
+        static RoutingKey createBindingKey(String sniHostname) {
+            if (sniHostname == null || sniHostname.isEmpty()) {
+                return NULL_ROUTING_KEY;
+            }
+            return new SniRoutingKey(sniHostname);
+        }
+
+    }
+
+    private static class NullRoutingKey implements RoutingKey {
+        @Override
+        public String toString() {
+            return "NullRoutingKey[]";
+        }
+    }
+
+    private record SniRoutingKey(String sniHostname) implements RoutingKey {
+
+    private SniRoutingKey(String sniHostname) {
+            Objects.requireNonNull(sniHostname);
+            this.sniHostname = sniHostname.toLowerCase(Locale.ROOT);
+        }}
+
+    protected static final AttributeKey<Map<RoutingKey, VirtualClusterBinding>> CHANNEL_BINDINGS = AttributeKey.newInstance("channelBindings");
+
+    private record VirtualClusterRecord(CompletionStage<Endpoint> registrationStage, AtomicReference<CompletionStage<Void>> deregistrationStage) {
+
+    private static VirtualClusterRecord create(CompletionStage<Endpoint> stage) {
+            return new VirtualClusterRecord(stage, new AtomicReference<>());
+        }}
+
+    /** Queue of network binding operations */
+    private final BlockingQueue<NetworkBindingOperation<?>> queue = new LinkedBlockingQueue<>();
+
+    /** Registry of virtual clusters that have been registered */
+    private final Map<VirtualCluster, VirtualClusterRecord> registeredVirtualClusters = new ConcurrentHashMap<>();
```

Comment:
Yeah, that's true.  The implementation is relying on the fact that the object model is created once as the configuration is parsed.  This will need to change as soon as we are capable of reloading configuration.

Are you happy if we defer this work?